src/transform/src/predicate_pushdown.rs

```diff
@@ -607,6 +619,69 @@ impl PredicatePushdown {
         }
     }
 
+    /// If `s` is of the form
+    /// `(isnull(expr1) && isnull(expr2)) || (expr1 = expr2)`,
+    /// extract `expr1` and `expr2`.
+    fn extract_equal_or_both_null(s: &mut MirScalarExpr) -> Option<(MirScalarExpr, MirScalarExpr)> {
+        // Or, And, and Eq are all commutative functions. For each of these
```

Comment:
I think this is another example of the issue above, but this one has an actual join condition:

```
materialize=> EXPLAIN SELECT * FROM t1 RIGHT JOIN t2 ON ( t1.f2 < t2.f1 ) WHERE t2.f1 + t1.f2 = 6 OR 2 > t2.f2;
                  Optimized Plan                   
---------------------------------------------------
 %0 =                                             +
 | Get materialize.public.t1 (u45)                +
 | ArrangeBy ()                                   +
                                                  +
 %1 =                                             +
 | Get materialize.public.t2 (u47)                +
                                                  +
 %2 = Let l0 =                                    +
 | Join %0 %1                                     +
 | | implementation = Differential %1 %0.()       +
 | | demand = (#0..#3)                            +
 | Filter ((6 = (#2 + #1)) || (2 > #3)), (#1 < #2)+ Filter #1, with join condition
                                                  +
 %3 =                                             +
 | Get %2 (l0)                                    +
 | Filter ((6 = (#2 + #1)) || (2 > #3))           + Filter #2, without join condition
```